### ISSUE TITLE
Cannot set Primary DHCP Server

### ISSUE THREAD

**Issue by keestux** [jefferyharrell/IPA-dhcp#2]
In the configuration of DHCP I am not able to set "Primary Server".

However, when I added the a host in the DHCP -> Servers dialog (the host being the IPA server that is also the actual DHCP server) I see that host in the list of "Secondary Servers".

The DHCP server fails to start giving the following error (with my own domain instead of example.com):

    Internet Systems Consortium DHCP Server 4.3.3
    Copyright 2004-2015 Internet Systems Consortium.
    All rights reserved.
    For info, please visit https://www.isc.org/software/dhcp/
    Error: Cannot find dhcpService DN 'cn=dhcp,dc=example,dc=com' with server reference. Please update the LDAP server entry 'cn=rotte.example.com,cn=dhcp,dc=example,dc=com'
    Configuration file errors encountered -- exiting

The dhcpd.conf simply has these ldap lines:

    ldap-server "rotte.example.com";
    ldap-port 389;
    ldap-base-dn "cn=dhcp,dc=example,dc=com";
    ldap-method static;
    ldap-debug-file "/var/log/dhcp-ldap-startup.log";

The relevant part of ldapsearch output is:

    # dhcp, example.com
    dn: cn=dhcp,dc=example,dc=com
    objectClass: top
    objectClass: dhcpService
    dhcpStatements: authoritative
    dhcpStatements: default-lease-time 43200
    dhcpStatements: max-lease-time 86400
    dhcpStatements: one-lease-per-client on
    cn: dhcp
    dhcpOption: domain-name "example.com"
    dhcpOption: domain-search "example.com"
    dhcpOption: domain-name-servers rotte.example.com
    dhcpSecondaryDN: cn=rotte.example.com,cn=dhcp,dc=example,dc=com

    # rotte.example.com, dhcp, example.com
    dn: cn=rotte.example.com,cn=dhcp,dc=example,dc=com
    objectClass: dhcpserver
    objectClass: top
    dhcpServiceDN: cn=dhcp,dc=example,dc=com
    cn: rotte.example.com

So, my dhcpserver is there, only it is the dhcpSecondaryDN. Maybe the dhcp server wants to see a primary server?

**Comment by keestux** [comment]
Hmm. It's a permission problem with LDAP attributes.

TL;DR
The dhcp server queries the LDAP server with something like (this is from the dhcp server source code):

    (&(objectClass=dhcpService)(|(|(dhcpPrimaryDN=%s)(dhcpSecondaryDN=%s))(dhcpServerDN=%s)))

Now apparently, in my case with my freeipa/ldap server, the dhcpServerDN is not a public attribute. And because dhcpServerDN isn't available for anonymous queries the request always fails. Even if dhcpSecondaryDN is present.

**Comment by keestux** [comment]
Found the (a) solution. In FreeIPA I had to add "dhcpserverdn" to the Effective attributes of the permission "System: Read DHCP Configuration".

Here are the steps I followed:

- login as admin to FreeIPA. Go to "IPA Server" -> "Role Based Access Control" -> "Permissions"
- search "dhcp" to find "System: Read DHCP Configuration"
- add "dhcpserverdn" to "Effective attributes"
- Save

BTW. I don't need to have a Primary Server (as I asked in the subject of this issue)
